File: java/src/main/java/com/example/minio/BasicOperations.java
package com.example.minio;

import io.minio.*;
import io.minio.messages.Item;
import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;

/**
 * Exemplos básicos de operações com MinIO usando Java.
 * Demonstra operações fundamentais como criar buckets, upload, download e listagem.
 */
public class BasicOperations {

    private static final String ENDPOINT = "http://localhost:9000";
    private static final String ACCESS_KEY = "minioadmin";
    private static final String SECRET_KEY = "minioadmin";

    private static MinioClient minioClient;

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("MinIO Java - Operações Básicas");
        System.out.println("=".repeat(60));

        try {
            // Inicializar cliente MinIO
            initializeClient();

            String bucketName = "java-example-bucket";

            // 1. Criar bucket
            System.out.println("\n1. Criando bucket...");
            createBucket(bucketName);

            // 2. Upload de string
            System.out.println("\n2. Fazendo upload de conteúdo...");
            uploadString(bucketName, "hello.txt", "Olá MinIO! Este é um teste com Java.");

            // 3. Upload de arquivo
            System.out.println("\n3. Criando e fazendo upload de arquivo...");
            String tempFile = "/tmp/java-example.txt";
            String content = "Este é um arquivo de exemplo criado em Java.\n" +
                           "MinIO é compatível com Amazon S3.\n";
            Files.write(Paths.get(tempFile), content.getBytes());
            uploadFile(bucketName, "example.txt", tempFile);

            // 4. Listar objetos
            System.out.println("\n4. Listando objetos...");
            listObjects(bucketName);

            // 5. Obter informações do objeto
            System.out.println("\n5. Obtendo informações do objeto...");
            getObjectInfo(bucketName, "example.txt");

            // 6. Download de arquivo
            System.out.println("\n6. Fazendo download...");
            String downloadPath = "/tmp/downloaded-java-example.txt";
            downloadFile(bucketName, "example.txt", downloadPath);

            String downloadedContent = new String(Files.readAllBytes(Paths.get(downloadPath)));
            System.out.println("   Conteúdo baixado:");
            System.out.println("   " + downloadedContent);

            // 7. Deletar objeto (opcional)
            // System.out.println("\n7. Removendo objeto...");
            // deleteObject(bucketName, "hello.txt");

            System.out.println("\n" + "=".repeat(60));
            System.out.println("✓ Todos os exemplos foram executados!");
            System.out.println("=".repeat(60));

        } catch (Exception e) {
            System.err.println("✗ Erro na execução: " + e.getMessage());
            e.printStackTrace();
        }
    }

    /**
     * Inicializa o cliente MinIO
     */
    private static void initializeClient() {
        minioClient = MinioClient.builder()
                .endpoint(ENDPOINT)
                .credentials(ACCESS_KEY, SECRET_KEY)
                .build();
        System.out.println("✓ Cliente MinIO inicializado");
    }

    /**
     * Cria um novo bucket
     */
    private static void createBucket(String bucketName) {
        try {
            boolean exists = minioClient.bucketExists(
                    BucketExistsArgs.builder().bucket(bucketName).build()
            );

            if (!exists) {
                minioClient.makeBucket(
                        MakeBucketArgs.builder().bucket(bucketName).build()
                );
                System.out.println("✓ Bucket '" + bucketName + "' criado com sucesso!");
            } else {
                System.out.println("✓ Bucket '" + bucketName + "' já existe.");
            }
        } catch (Exception e) {
            System.err.println("✗ Erro ao criar bucket: " + e.getMessage());
        }
    }

    /**
     * Faz upload de uma string como objeto
     */
    private static void uploadString(String bucketName, String objectName, String content) {
        try {
            byte[] bytes = content.getBytes();
            ByteArrayInputStream bais = new ByteArrayInputStream(bytes);

            minioClient.putObject(
                    PutObjectArgs.builder()
                            .bucket(bucketName)
                            .object(objectName)
                            .stream(bais, bytes.length, -1)
                            .contentType("text/plain")
                            .build()
            );

            System.out.println("✓ Conteúdo enviado como '" + objectName + "'");
        } catch (Exception e) {
            System.err.println("✗ Erro no upload: " + e.getMessage());
        }
    }

    /**
     * Faz upload de um arquivo
     */
    private static void uploadFile(String bucketName, String objectName, String filePath) {
        try {
            minioClient.uploadObject(
                    UploadObjectArgs.builder()
                            .bucket(bucketName)
                            .object(objectName)
                            .filename(filePath)
                            .contentType("text/plain")
                            .build()
            );

            System.out.println("✓ Arquivo '" + filePath + "' enviado como '" + objectName + "'");
        } catch (Exception e) {
            System.err.println("✗ Erro no upload: " + e.getMessage());
        }
    }

    /**
     * Lista todos os objetos em um bucket
     */
    private static void listObjects(String bucketName) {
        try {
            System.out.println("\n📋 Objetos no bucket '" + bucketName + "':");
            
            Iterable<Result<Item>> results = minioClient.listObjects(
                    ListObjectsArgs.builder()
                            .bucket(bucketName)
                            .build()
            );

            for (Result<Item> result : results) {
                Item item = result.get();
                System.out.println("  - " + item.objectName() + 
                                 " (Tamanho: " + item.size() + " bytes)");
            }
        } catch (Exception e) {
            System.err.println("✗ Erro ao listar objetos: " + e.getMessage());
        }
    }

    /**
     * Obtém informações sobre um objeto
     */
    private static void getObjectInfo(String bucketName, String objectName) {
        try {
            StatObjectResponse stat = minioClient.statObject(
                    StatObjectArgs.builder()
                            .bucket(bucketName)
                            .object(objectName)
                            .build()
            );

            System.out.println("\n📊 Informações do objeto '" + objectName + "':");
            System.out.println("  - Tamanho: " + stat.size() + " bytes");
            System.out.println("  - Tipo: " + stat.contentType());
            System.out.println("  - Última modificação: " + stat.lastModified());
            System.out.println("  - ETag: " + stat.etag());
        } catch (Exception e) {
            System.err.println("✗ Erro ao obter informações: " + e.getMessage());
        }
    }

    /**
     * Faz download de um arquivo
     */
    private static void downloadFile(String bucketName, String objectName, String filePath) {
        try {
            minioClient.downloadObject(
                    DownloadObjectArgs.builder()
                            .bucket(bucketName)
                            .object(objectName)
                            .filename(filePath)
                            .build()
            );

            System.out.println("✓ Objeto '" + objectName + "' baixado para '" + filePath + "'");
        } catch (Exception e) {
            System.err.println("✗ Erro no download: " + e.getMessage());
        }
    }

    /**
     * Remove um objeto do bucket
     */
    private static void deleteObject(String bucketName, String objectName) {
        try {
            minioClient.removeObject(
                    RemoveObjectArgs.builder()
                            .bucket(bucketName)
                            .object(objectName)
                            .build()
            );

            System.out.println("✓ Objeto '" + objectName + "' removido");
        } catch (Exception e) {
            System.err.println("✗ Erro ao remover objeto: " + e.getMessage());
        }
    }
}

File: go/basic_operations.go
package main

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"log"
	"os"
	"strings"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
)

const (
	endpoint  = "localhost:9000"
	accessKey = "minioadmin"
	secretKey = "minioadmin"
	useSSL    = false
)

// Operações básicas com MinIO usando Go
func main() {
	fmt.Println(strings.Repeat("=", 60))
	fmt.Println("MinIO Go - Operações Básicas")
	fmt.Println(strings.Repeat("=", 60))

	// Inicializar cliente MinIO
	client, err := minio.New(endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(accessKey, secretKey, ""),
		Secure: useSSL,
	})
	if err != nil {
		log.Fatalf("✗ Erro ao inicializar cliente: %v", err)
	}
	fmt.Println("✓ Cliente MinIO inicializado")

	ctx := context.Background()
	bucketName := "go-example-bucket"

	// 1. Criar bucket
	fmt.Println("\n1. Criando bucket...")
	if err := createBucket(ctx, client, bucketName); err != nil {
		log.Printf("✗ Erro: %v", err)
	}

	// 2. Upload de string
	fmt.Println("\n2. Fazendo upload de conteúdo...")
	content := "Olá MinIO! Este é um teste com Go."
	if err := uploadString(ctx, client, bucketName, "hello.txt", content); err != nil {
		log.Printf("✗ Erro: %v", err)
	}

	// 3. Upload de arquivo
	fmt.Println("\n3. Criando e fazendo upload de arquivo...")
	tempFile := "/tmp/go-example.txt"
	fileContent := "Este é um arquivo de exemplo criado em Go.\nMinIO é compatível com Amazon S3.\n"
	if err := os.WriteFile(tempFile, []byte(fileContent), 0644); err != nil {
		log.Printf("✗ Erro ao criar arquivo: %v", err)
	} else {
		if err := uploadFile(ctx, client, bucketName, "example.txt", tempFile); err != nil {
			log.Printf("✗ Erro: %v", err)
		}
	}

	// 4. Listar objetos
	fmt.Println("\n4. Listando objetos...")
	if err := listObjects(ctx, client, bucketName); err != nil {
		log.Printf("✗ Erro: %v", err)
	}

	// 5. Obter informações do objeto
	fmt.Println("\n5. Obtendo informações do objeto...")
	if err := getObjectInfo(ctx, client, bucketName, "example.txt"); err != nil {
		log.Printf("✗ Erro: %v", err)
	}

	// 6. Download de arquivo
	fmt.Println("\n6. Fazendo download...")
	downloadPath := "/tmp/downloaded-go-example.txt"
	if err := downloadFile(ctx, client, bucketName, "example.txt", downloadPath); err != nil {
		log.Printf("✗ Erro: %v", err)
	} else {
		downloadedContent, _ := os.ReadFile(downloadPath)
		fmt.Printf("   Conteúdo baixado:\n   %s", string(downloadedContent))
	}

	// 7. Deletar objeto (opcional)
	// fmt.Println("\n7. Removendo objeto...")
	// deleteObject(ctx, client, bucketName, "hello.txt")

	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("✓ Todos os exemplos foram executados!")
	fmt.Println(strings.Repeat("=", 60))
}

// createBucket cria um novo bucket
func createBucket(ctx context.Context, client *minio.Client, bucketName string) error {
	exists, err := client.BucketExists(ctx, bucketName)
	if err != nil {
		return fmt.Errorf("erro ao verificar bucket: %w", err)
	}

	if !exists {
		err = client.MakeBucket(ctx, bucketName, minio.MakeBucketOptions{})
		if err != nil {
			return fmt.Errorf("erro ao criar bucket: %w", err)
		}
		fmt.Printf("✓ Bucket '%s' criado com sucesso!\n", bucketName)
	} else {
		fmt.Printf("✓ Bucket '%s' já existe.\n", bucketName)
	}
	return nil
}

// uploadString faz upload de uma string como objeto
func uploadString(ctx context.Context, client *minio.Client, bucketName, objectName, content string) error {
	reader := bytes.NewReader([]byte(content))
	
	_, err := client.PutObject(ctx, bucketName, objectName, reader, int64(len(content)),
		minio.PutObjectOptions{ContentType: "text/plain"})
	if err != nil {
		return fmt.Errorf("erro no upload: %w", err)
	}

	fmt.Printf("✓ Conteúdo enviado como '%s'\n", objectName)
	return nil
}

// uploadFile faz upload de um arquivo
func uploadFile(ctx context.Context, client *minio.Client, bucketName, objectName, filePath string) error {
	_, err := client.FPutObject(ctx, bucketName, objectName, filePath,
		minio.PutObjectOptions{ContentType: "text/plain"})
	if err != nil {
		return fmt.Errorf("erro no upload: %w", err)
	}

	fmt.Printf("✓ Arquivo '%s' enviado como '%s'\n", filePath, objectName)
	return nil
}

// listObjects lista todos os objetos em um bucket
func listObjects(ctx context.Context, client *minio.Client, bucketName string) error {
	fmt.Printf("\n📋 Objetos no bucket '%s':\n", bucketName)

	objectCh := client.ListObjects(ctx, bucketName, minio.ListObjectsOptions{})
	for object := range objectCh {
		if object.Err != nil {
			return fmt.Errorf("erro ao listar objetos: %w", object.Err)
		}
		fmt.Printf("  - %s (Tamanho: %d bytes)\n", object.Key, object.Size)
	}
	return nil
}

// getObjectInfo obtém informações sobre um objeto
func getObjectInfo(ctx context.Context, client *minio.Client, bucketName, objectName string) error {
	objInfo, err := client.StatObject(ctx, bucketName, objectName, minio.StatObjectOptions{})
	if err != nil {
		return fmt.Errorf("erro ao obter informações: %w", err)
	}

	fmt.Printf("\n📊 Informações do objeto '%s':\n", objectName)
	fmt.Printf("  - Tamanho: %d bytes\n", objInfo.Size)
	fmt.Printf("  - Tipo: %s\n", objInfo.ContentType)
	fmt.Printf("  - Última modificação: %s\n", objInfo.LastModified)
	fmt.Printf("  - ETag: %s\n", objInfo.ETag)
	return nil
}

// downloadFile faz download de um arquivo
func downloadFile(ctx context.Context, client *minio.Client, bucketName, objectName, filePath string) error {
	err := client.FGetObject(ctx, bucketName, objectName, filePath, minio.GetObjectOptions{})
	if err != nil {
		return fmt.Errorf("erro no download: %w", err)
	}

	fmt.Printf("✓ Objeto '%s' baixado para '%s'\n", objectName, filePath)
	return nil
}

// deleteObject remove um objeto do bucket
func deleteObject(ctx context.Context, client *minio.Client, bucketName, objectName string) error {
	err := client.RemoveObject(ctx, bucketName, objectName, minio.RemoveObjectOptions{})
	if err != nil {
		return fmt.Errorf("erro ao remover objeto: %w", err)
	}

	fmt.Printf("✓ Objeto '%s' removido\n", objectName)
	return nil
}

// downloadToMemory baixa um objeto para memória
func downloadToMemory(ctx context.Context, client *minio.Client, bucketName, objectName string) ([]byte, error) {
	object, err := client.GetObject(ctx, bucketName, objectName, minio.GetObjectOptions{})
	if err != nil {
		return nil, fmt.Errorf("erro ao obter objeto: %w", err)
	}
	defer object.Close()

	data, err := io.ReadAll(object)
	if err != nil {
		return nil, fmt.Errorf("erro ao ler objeto: %w", err)
	}

	return data, nil
}
